controllers/virtualmachine/virtualmachine_controller.go

```diff
@@ -234,9 +226,13 @@ func (r *Reconciler) Reconcile(ctx context.Context, req ctrl.Request) (_ ctrl.Re
 		}
 	}()
 
+	// When PausedVMLabelKey exists, VM is being paused from deletion.
 	if !vm.DeletionTimestamp.IsZero() {
-		err = r.ReconcileDelete(vmCtx)
-		return ctrl.Result{}, err
+		if _, exists := vm.Labels[vmopv1.PausedVMLabelKey]; !exists {
+			err = r.ReconcileDelete(vmCtx)
+			return ctrl.Result{}, err
+		}
+		vmCtx.Logger.Info("ReconcileDelete not called because VirtualMachine is currently paused")
```

Comment:
This is missing a return.

Can we move this check into `ReconcileDelete()` to keep Reconcile simpler and making this easier to unit test?

Reply:
Didn't return on purpose. So that when label is changed, it can be reconciled to update.
Moved to `ReconcileDelete()`, checking for PauseAnnotation instead of label. So that update on admin removing EC can be reconciled.
```
func (r *Reconciler) ReconcileDelete(ctx *pkgctx.VirtualMachineContext) (reterr error) {
	ctx.Logger.Info("Reconciling VirtualMachine Deletion")

	// If the VM reconciliation has been paused by the developer,
	// skip deletion and return.
	if _, exists := ctx.VM.Annotations[vmopv1.PauseAnnotation]; exists {
		ctx.Logger.Info("VirtualMachine is not deleted because devops paused this VM")
		return nil
	}

	if controllerutil.ContainsFinalizer(ctx.VM, finalizerName) {
		defer func() {
			r.Recorder.EmitEvent(ctx.VM, "Delete", reterr, false)
		}()

		if err := r.VMProvider.DeleteVirtualMachine(ctx, ctx.VM); err != nil {
			// If VM can not be deleted due to reconciliation being paused, ignore that.
			if errors.Is(err, vspherevm.ErrorVMPausedByAdmin()) {
				ctx.Logger.Info("VM could not be deleted since it contains the pause reconcile ExtraConfig key")
				return nil
			}
			ctx.Logger.Error(err, "Failed to delete VirtualMachine")
			return err
		}

		controllerutil.RemoveFinalizer(ctx.VM, finalizerName)
		ctx.Logger.Info("Provider Completed deleting Virtual Machine", "time", time.Now().Format(time.RFC3339))
	}

	// BMV: Shouldn't these be in the ContainsFinalizer block?
	r.vmMetrics.DeleteMetrics(ctx)
	r.Prober.RemoveFromProberManager(ctx.VM)

	ctx.Logger.Info("Finished Reconciling VirtualMachine Deletion")
	return nil
}
```